### ISSUE TITLE
Problem with upgradeStrategy and Scale

### ISSUE THREAD

**Issue by ch3lo** [mesosphere/marathon#3091]
Hi,

I have a problem deploying a Docker Service with Marathon **0.14** and Mesos **0.26.0**

I made a `PUT` to `localhost:8081/v2/apps` with the next payload and I have no problem. Marathon does exactly what I want.

``` json
[{
  "id": "/www/services/test-http/test-http-v0",
  "cpus": 0.1,
  "mem": 32,
  "instances": 5,
  "container": {
    "type": "DOCKER",
    "docker": {
      "image": "xxx/test-http:0-svn479-test-http",
      "network": "BRIDGE",
      "portMappings": [
        { "containerPort": 80, "hostPort": 0, "protocol": "tcp"}
      ]
    }
  },
  "healthChecks": [
    {
      "protocol": "HTTP",
      "portIndex": 0,
      "path": "/v0/healthy",
      "gracePeriodSeconds": 30,
      "intervalSeconds": 20,
      "maxConsecutiveFailures": 3
    }
  ],
  "upgradeStrategy": {
    "minimumHealthCapacity": 1,
    "maximumOverCapacity": 0.2
  }
}]
```

Then I made another `PUT` to `localhost:8081/v2/apps` with the objective to upgrade my service to `0-svn480-test-http` with this:

``` json
[{
  "id": "/www/services/test-http/test-http-v0",
  "cpus": 0.1,
  "mem": 32,
  "instances": 3,
  "container": {
    "type": "DOCKER",
    "docker": {
      "image": "xxx/test-http:0-svn480-test-http",
      "network": "BRIDGE",
      "portMappings": [
        { "containerPort": 80, "hostPort": 0, "protocol": "tcp"}
      ]
    }
  },
  "healthChecks": [
    {
      "protocol": "HTTP",
      "portIndex": 0,
      "path": "/v0/healthy",
      "gracePeriodSeconds": 30,
      "intervalSeconds": 20,
      "maxConsecutiveFailures": 3
    }
  ],
  "upgradeStrategy": {
    "minimumHealthCapacity": 1,
    "maximumOverCapacity": 0.2
  }
}]
```

With `minimumHealthCapacity` **1** Marathon scales `0-svn479-test-http` from **5** to **3** (I think it is ok, because my scale parameter is 3). The next step should be deploy the new service (one instance at time, because my maximumOverCapacity is 0.2), however Marathon **hangs** in **RestartApplication**.

NOTE: If I use a scale value greater or equal, the deploy works well.

Any suggestion?

Regards

**Comment by Poltergeist** [comment]
Hi @ch3lo,
I will look into that, but I got one question, did you try to change the docker image, wait till it is deployed and then scale down the application? 
This could be a valid workaround.

Regards

**Comment by ch3lo** [comment]
Hi @Poltergeist , we changed the image name but only the tag part. Marathon scales down but it holds a lot of time without deploying a new instance.

Regards

**Comment by Poltergeist** [comment]
I reproduced this issue. I solved it by doing two steps instead of one. I first changed the docker image tag and then in another request the instance size. 

I also did another test were I tried to solve it in one request. I changed the `maximumOverCapacity` to `0.4`, that worked. So the problem might be that the `maximumOverCapacity` is a factor of the new `instances` size which while scaling down is `3 * 0.2 = 0,6`. So the `maximumOverCapacity` should default to a minimum of 1. This also has nothing to do with a docker image it can also be reproduced with any other application.

So the best solution for now would be to take care that the `maximumOverCapacity * instances` is bigger or equal to 1.

**Comment by ch3lo** [comment]
Thank you. I will try with your alternatives but I think  maximumOverCapacity * instances < 1 should return a error before to upgrade.

Regards

**Comment by pgkelley4** [comment]
@Poltergeist we would like to start using the `maximumOverCapacity` feature (i.e. setting it to something other than 1) but this bug makes it tough. I narrowed down what I think the problem is and put up a pull request here: https://github.com/mesosphere/marathon/pull/3531

**Comment by tomwganem** [comment]
We've also ran into this bug and can't seem to get the workaround working. Would it be possible to include the PR that fixes this bug into a 0.15.4 release?

**Comment by meichstedt** [comment]
Note: This issue has been migrated to https://jira.mesosphere.com/browse/MARATHON-2685. For more information see https://groups.google.com/forum/#!topic/marathon-framework/khtvf-ifnp8.
